File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/common/PreAuthRecord.kt
package com.fccmiddleware.edge.adapter.common

import kotlinx.serialization.Serializable

/**
 * Domain model for a pre-authorization record.
 *
 * This is the canonical domain object — NOT the Room entity.
 * The Room entity lives in buffer/entity/PreAuthRecord.kt.
 *
 * Matches pre-auth-record.schema.json v1 field-for-field.
 * Money fields use Long minor units. Timestamps use ISO 8601 UTC strings.
 * UUIDs use String.
 *
 * Terminal states: COMPLETED, CANCELLED, EXPIRED, FAILED.
 */
@Serializable
data class PreAuthRecord(
    /** Middleware-generated UUID v4. Primary key. */
    val id: String,

    /** Site identifier. Part of dedup key with odooOrderId. */
    val siteCode: String,

    /** Odoo POS order reference. Part of dedup key with siteCode. */
    val odooOrderId: String,

    /** Target pump number. */
    val pumpNumber: Int,

    /** Target nozzle number. */
    val nozzleNumber: Int,

    /** Canonical fuel product code (e.g., PMS, AGO, IK, DPK). */
    val productCode: String,

    /** Authorized amount in integer minor units (cents). Always > 0. */
    val requestedAmount: Long,

    /** Price per litre at time of authorization, in minor units per litre. */
    val unitPrice: Long,

    /** ISO 4217 currency code (e.g., ZAR, KES, TZS). */
    val currency: String,

    /** Current pre-authorization lifecycle state. */
    val status: PreAuthStatus,

    /** When Odoo POS submitted the pre-auth request to the Edge Agent. UTC ISO 8601. */
    val requestedAt: String,

    /** requestedAt + preAuthTimeoutMinutes. UTC ISO 8601. */
    val expiresAt: String,

    /** Record creation timestamp in UTC ISO 8601. */
    val createdAt: String,

    /** Last modification timestamp. Updated on every state transition. UTC ISO 8601. */
    val updatedAt: String,

    /** Schema version for forward compatibility. Current value: 1. */
    val schemaVersion: Int,

    /** Vehicle registration plate, if captured. */
    val vehicleNumber: String? = null,

    /** Customer name. Required when requireCustomerTaxId=true. */
    val customerName: String? = null,

    /** Customer Tax Identification Number (TIN). PII — NEVER log. */
    val customerTaxId: String? = null,

    /** B2B business name. Required for B2B fiscalized transactions. */
    val customerBusinessName: String? = null,

    /** Odoo user ID of the fuel attendant. */
    val attendantId: String? = null,

    /** FCC-assigned correlation reference returned at authorization time. */
    val fccCorrelationId: String? = null,

    /** FCC authorization confirmation code. Null until AUTHORIZED state. */
    val fccAuthorizationCode: String? = null,

    /** fccTransactionId of the final dispense matched by the reconciliation engine. Null until COMPLETED. */
    val matchedFccTransactionId: String? = null,

    /** Actual dispensed amount in integer minor units. Null until COMPLETED. */
    val actualAmount: Long? = null,

    /** Actual dispensed volume in integer millilitres (1 litre = 1000). Null until COMPLETED. */
    val actualVolume: Long? = null,

    /** actualAmount - requestedAmount in minor units. Null on Edge Agent. */
    val amountVariance: Long? = null,

    /** ABS(amountVariance) / requestedAmount * 10000, rounded (basis points). Null on Edge Agent. */
    val varianceBps: Long? = null,

    /** When FCC returned authorization confirmation. Null until AUTHORIZED. UTC ISO 8601. */
    val authorizedAt: String? = null,

    /** When FCC signalled dispense started. Null if vendor does not report DISPENSING. UTC ISO 8601. */
    val dispensingAt: String? = null,

    /** When reconciliation engine matched the final dispense. Null until COMPLETED. UTC ISO 8601. */
    val completedAt: String? = null,

    /** When cancellation was processed. Null unless CANCELLED. UTC ISO 8601. */
    val cancelledAt: String? = null,

    /** When expiry worker transitioned the record. Null unless EXPIRED. UTC ISO 8601. */
    val expiredAt: String? = null,

    /** When FCC rejection was processed. Null unless FAILED. UTC ISO 8601. */
    val failedAt: String? = null,
)

File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Common/PreAuthRecord.cs
using System.Text.Json.Serialization;
using FccDesktopAgent.Core.Security;

namespace FccDesktopAgent.Core.Adapter.Common;

/// <summary>
/// Domain model for a pre-authorization record.
/// Not an EF entity — used for in-memory processing and cloud serialization.
/// Idempotency key: (OdooOrderId, SiteCode).
/// </summary>
public sealed record PreAuthRecord
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = Guid.NewGuid().ToString();

    [JsonPropertyName("siteCode")]
    public string SiteCode { get; init; } = string.Empty;

    [JsonPropertyName("odooOrderId")]
    public string OdooOrderId { get; init; } = string.Empty;

    [JsonPropertyName("pumpNumber")]
    public int PumpNumber { get; init; }

    [JsonPropertyName("nozzleNumber")]
    public int NozzleNumber { get; init; }

    [JsonPropertyName("productCode")]
    public string ProductCode { get; init; } = string.Empty;

    /// <summary>Requested authorization amount in minor currency units.</summary>
    [JsonPropertyName("requestedAmount")]
    public long RequestedAmount { get; init; }

    /// <summary>Unit price in minor units per litre at time of pre-auth.</summary>
    [JsonPropertyName("unitPrice")]
    public long UnitPrice { get; init; }

    [JsonPropertyName("currency")]
    public string Currency { get; init; } = string.Empty;

    [JsonPropertyName("status")]
    public PreAuthStatus Status { get; init; } = PreAuthStatus.Pending;

    [JsonPropertyName("vehicleNumber")]
    public string? VehicleNumber { get; init; }

    [SensitiveData, JsonPropertyName("customerName")]
    public string? CustomerName { get; init; }

    [SensitiveData, JsonPropertyName("customerTaxId")]
    public string? CustomerTaxId { get; init; }

    [JsonPropertyName("customerBusinessName")]
    public string? CustomerBusinessName { get; init; }

    [JsonPropertyName("attendantId")]
    public string? AttendantId { get; init; }

    [JsonPropertyName("fccCorrelationId")]
    public string? FccCorrelationId { get; init; }

    [JsonPropertyName("fccAuthorizationCode")]
    public string? FccAuthorizationCode { get; init; }

    [JsonPropertyName("matchedFccTransactionId")]
    public string? MatchedFccTransactionId { get; init; }

    [JsonPropertyName("actualAmount")]
    public long? ActualAmount { get; init; }

    [JsonPropertyName("actualVolume")]
    public long? ActualVolume { get; init; }

    [JsonPropertyName("amountVariance")]
    public long? AmountVariance { get; init; }

    [JsonPropertyName("varianceBps")]
    public int? VarianceBps { get; init; }

    [JsonPropertyName("requestedAt")]
    public DateTimeOffset RequestedAt { get; init; }

    [JsonPropertyName("expiresAt")]
    public DateTimeOffset ExpiresAt { get; init; }

    [JsonPropertyName("authorizedAt")]
    public DateTimeOffset? AuthorizedAt { get; init; }

    [JsonPropertyName("dispensingAt")]
    public DateTimeOffset? DispensingAt { get; init; }

    [JsonPropertyName("completedAt")]
    public DateTimeOffset? CompletedAt { get; init; }

    [JsonPropertyName("cancelledAt")]
    public DateTimeOffset? CancelledAt { get; init; }

    [JsonPropertyName("expiredAt")]
    public DateTimeOffset? ExpiredAt { get; init; }

    [JsonPropertyName("failedAt")]
    public DateTimeOffset? FailedAt { get; init; }

    [JsonPropertyName("schemaVersion")]
    public string SchemaVersion { get; init; } = "1.0";

    [JsonPropertyName("createdAt")]
    public DateTimeOffset CreatedAt { get; init; }

    [JsonPropertyName("updatedAt")]
    public DateTimeOffset UpdatedAt { get; init; }
}
